mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1374,6 +1375,63 @@ pub fn create_taker_spends_maker_payment_preimage<T: UtxoCommonOps + SwapOps>(
     Box::new(fut.boxed().compat())
 }
 
+pub fn create_taker_refunds_payment<T: UtxoCommonOps + SwapOps>(
+    coin: T,
+    taker_payment_tx: &[u8],
+    time_lock: u32,
+    maker_pub: &[u8],
+    secret_hash: &[u8],
+    swap_unique_data: &[u8],
+) -> TransactionFut {
+    let my_address = try_tx_fus!(coin.as_ref().derivation_method.iguana_or_err()).clone();
+    let mut prev_transaction: UtxoTx =
+        try_tx_fus!(deserialize(taker_payment_tx).map_err(|e| TransactionErr::Plain(format!("{:?}", e))));
+    prev_transaction.tx_hash_algo = coin.as_ref().tx_hash_algo;
+    drop_mutability!(prev_transaction);
+    if prev_transaction.outputs.is_empty() {
+        return try_tx_fus!(TX_PLAIN_ERR!("Transaction doesn't have any output"));
+    }
+
+    let key_pair = coin.derive_htlc_key_pair(swap_unique_data);
+    let script_data = Builder::default().push_opcode(Opcode::OP_1).into_script();
+    let redeem_script = payment_script(
+        time_lock,
+        secret_hash,
+        key_pair.public(),
+        &try_tx_fus!(Public::from_slice(maker_pub)),
+    )
+    .into();
+    let fut = async move {
+        let fee = try_tx_s!(coin.get_htlc_spend_fee(DEFAULT_SWAP_TX_SPEND_SIZE).await);
+        if fee >= prev_transaction.outputs[0].value {
+            return TX_PLAIN_ERR!(
+                "HTLC spend fee {} is greater than transaction output {}",
+                fee,
+                prev_transaction.outputs[0].value
+            );
+        }
+        let script_pubkey = output_script(&my_address, ScriptType::P2PKH).to_bytes();
+        let output = TransactionOutput {
+            value: prev_transaction.outputs[0].value - fee,
+            script_pubkey,
+        };
+
+        let input = P2SHSpendingTxInput {
+            prev_transaction,
+            redeem_script,
+            outputs: vec![output],
+            script_data,
+            sequence: SEQUENCE_FINAL - 1,
```

Comment:
I also had to check it, it is explained [here](https://bitcoin.stackexchange.com/a/87376). Shortly, the relevant part is that setting the sequence number to `SEQUENCE_FINAL` disables the locktime. If it is set to `SEQUENCE_FINAL - 1` then the locktime is used. In our code, `taker_spends_maker_payment` and `maker_spends_taker_payment` use `SEQUENCE_FINAL` since the locktime is not used. Refund transactions use `SEQUENCE_FINAL - 1` since the locktime is used.

Reply:
Awesome, thanks! It's now clear :)